### ISSUE TITLE
RyuJIT/x86: CoreFx System.Reflection.Emit.Tests.dll Exit code = -2146233022

### ISSUE THREAD

**Issue by RussKeldorph** [dotnet/coreclr#8158]
**Repro steps:**

coreclr *(assuming root=c:\git\coreclr)*
```
build x86 checked skiptests
```
corefx *(assuming root=c:\git\corefx)*
```
build -Release -SkipTests -TestNugetRuntimeId=win7-x86 -- /p:BUILDTOOLS_OVERRIDE_RUNTIME=c:\git\coreclr\bin\Product\Windows_NT.x86.Checked
cd bin\tests\AnyOS.AnyCPU.Release\System.Reflection.Emit.Tests\default.netcoreapp1.1
set COMPlus_AltJit=*
set COMPlus_AltJitName=protojit.dll
RunTests.cmd c:\git\corefx\packages
```
**Output**
```
Using c:\git\corefx\packages as folder for resolving package dependencies.
Executing in C:\git\corefx\bin\tests\AnyOS.AnyCPU.Release\System.Reflection.Emit.Tests\default.netcoreapp1.1\
Hard linking dependent files...
Finished linking needed files, moving to running tests.
Running tests... Start time: 16:38:28.43
Command(s):
call CoreRun.exe xunit.console.netcore.exe System.Reflection.Emit.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapp1.1tests  -notrait category=OuterLoop -notrait category=failing -notrait category=nonwindowstests
xUnit.net console test runner (32-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Reflection.Emit.Tests
Discovered:  System.Reflection.Emit.Tests
Starting:    System.Reflection.Emit.Tests
Finished running tests.  End time=16:38:43.71, Exit code = -2146233022
```

**Comment by mikedn** [comment]
When run in debugger this shows an assert, not sure if it's related or a red herring:
```
Assert failure(PID 10348 [0x0000286c], Thread: 13580 [0x350c]): CONTRACT VIOLATION by ReflectionInvocationExceptionFilter at "d:\projects\coreclr\src\vm\excep.cpp" @ 8937

THROWS called in a NOTHROW region.

                        CONTRACT in ReflectionInvocationExceptionFilter at "d:\projects\coreclr\src\vm\excep.cpp" @ 8937
VIOLATED-->  CONTRACT in SigPointer::GetTypeVariable at "d:\projects\coreclr\src\vm\siginfo.cpp" @ 1888
                        CONTRACT in SigPointer::GetTypeVariableThrowing at "d:\projects\coreclr\src\vm\siginfo.cpp" @ 1858
                        CONTRACT in SigPointer::GetTypeHandleThrowing at "d:\projects\coreclr\src\vm\siginfo.cpp" @ 1007
                        CONTRACT in SigPointer::GetTypeHandleThrowing at "d:\projects\coreclr\src\vm\siginfo.cpp" @ 1007
                        CONTRACT in ClassLoader::LoadApproxTypeThrowing at "d:\projects\coreclr\src\vm\clsload.cpp" @ 3354
                        CONTRACT in ClassLoader::LoadApproxParentThrowing at "d:\projects\coreclr\src\vm\clsload.cpp" @ 3451
                        OVERRIDE_TYPE_LOAD_LEVEL_LIMIT in ClassLoader::CreateTypeHandleForTypeDefThrowing at "d:\projects\coreclr\src\vm\methodtablebuilder.cpp" @ 13166
                        CONTRACT in ClassLoader::CreateTypeHandleForTypeDefThrowing at "d:\projects\coreclr\src\vm\methodtablebuilder.cpp" @ 13117
                        CONTRACT in ClassLoader::CreateTypeHandleForTypeKey at "d:\projects\coreclr\src\vm\clsload.cpp" @ 3598
                        CONTRACT in ClassLoader::DoIncrementalLoad at "d:\projects\coreclr\src\vm\clsload.cpp" @ 3511
                        CONTRACT in ClassLoader::LoadTypeHandleForTypeKey_Body at "d:\projects\coreclr\src\vm\clsload.cpp" @ 4167
                        GCX_PREEMP in ClassLoader::LoadTypeHandleForTypeKey at "d:\projects\coreclr\src\vm\clsload.cpp" @ 4018
                        CONTRACT in ClassLoader::LoadTypeHandleForTypeKey at "d:\projects\coreclr\src\vm\clsload.cpp" @ 4016
                        CONTRACT in COMDynamicWrite::TermCreateClass at "d:\projects\coreclr\src\vm\comdynamic.cpp" @ 462
                        OVERRIDE_TYPE_LOAD_LEVEL_LIMIT in MethodDescCallSite::CallTargetWorker at "d:\projects\coreclr\src\vm\callhelpers.cpp" @ 382
                        CONTRACT in MethodDescCallSite::CallTargetWorker at "d:\projects\coreclr\src\vm\callhelpers.cpp" @ 372
                        CONTRACT in ThreadNative::KickOffThread_Worker at "d:\projects\coreclr\src\vm\comsynchronizable.cpp" @ 225
                        CONTRACT in ManagedThreadBase_DispatchInner at "d:\projects\coreclr\src\vm\threads.cpp" @ 9995
                        CONTRACT in ManagedThreadBase_FullTransitionWithAD at "d:\projects\coreclr\src\vm\threads.cpp" @ 10396
                        ...



CORECLR! CONTRACT_ASSERT + 0x325 (0x5ee18760)
CORECLR! ClrDebugState::CheckOkayToThrow + 0x23 (0x5ee18bf4)
CORECLR! BaseContract::DoChecks + 0x125 (0x5ee19ee6)
CORECLR! EEContract::DoChecks + 0x5F (0x5eedada2)
CORECLR! ReflectionInvocationExceptionFilter + 0x75 (0x5ef53d6a)
CORECLR! CallDescrWorkerReflectionWrapper + 0x4A (0x5f355047)
CORECLR! EH4_CallFilterFunc + 0x12 (0x5f510240)
CORECLR! except_handler4 + 0xC1 (0x5f50d041)
NTDLL! RtlConvertUlongToLargeInteger + 0x122 (0x77201952)
NTDLL! RtlConvertUlongToLargeInteger + 0xF4 (0x77201924)
    File: d:\projects\coreclr\src\vm\excep.cpp Line: 8937
    Image: D:\Projects\corefx\bin\tests\AnyOS.AnyCPU.Release\System.Reflection.Emit.Tests\default.netcoreapp1.1\CoreRun.exe

Finished running tests.  End time= 0:43:02.83, Exit code = 123456789
```

**Comment by pgavlin** [comment]
@mikedn: that failure does correspond with the HRESULT that is being returned as the exit code for corerun.

@jkotas: this looks like a runtime issue. Who should we route this to?

**Comment by jkotas** [comment]
I see two different failures modes with CoreCLR debug x86:

- Deterministic failure with -2146233022 exit code. It is duplicate of https://github.com/dotnet/corefx/issues/13238:

```
CoreCLR!SystemNative::Exit+0x180 [c:\coreclr\src\classlibnative\bcltype\system.cpp @ 103]
System_Private_CoreLib!DomainNeutralILStubClass.IL_STUB_PInvoke(Int32)+0x75
System_Private_CoreLib!System.Diagnostics.Assert.Fail(System.String, System.String, System.String, Int32, TraceFormat, Int32)+0x16d [C:\coreclr\src\mscorlib\src\System\Diagnostics\Assert.cs @ 89]
System_Private_CoreLib!System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation(System.Diagnostics.Contracts.ContractFailureKind, System.String, System.String, System.String, System.Exception)+0xb3 [C:\coreclr\src\mscorlib\src\System\Diagnostics\Contracts\ContractsBCL.cs @ 425]
System_Private_CoreLib!System.Runtime.CompilerServices.ContractHelper.TriggerFailure(System.Diagnostics.Contracts.ContractFailureKind, System.String, System.String, System.String, System.Exception)+0x30 [C:\coreclr\src\mscorlib\src\System\Diagnostics\Contracts\Contracts.cs @ 989]
System_Private_CoreLib!System.Diagnostics.Contracts.Contract.ReportFailure(System.Diagnostics.Contracts.ContractFailureKind, System.String, System.String, System.Exception)+0x11c [C:\coreclr\src\mscorlib\src\System\Diagnostics\Contracts\ContractsBCL.cs @ 110]
System_Private_CoreLib!System.Diagnostics.Contracts.Contract.Assert(Boolean)+0x45 [C:\coreclr\src\mscorlib\src\System\Diagnostics\Contracts\Contracts.cs @ 343]
System_Private_CoreLib!System.Reflection.Emit.ModuleBuilder.DefineEnum(System.String, System.Reflection.TypeAttributes, System.Type)+0xd5 [C:\coreclr\src\mscorlib\src\System\Reflection\Emit\ModuleBuilder.cs @ 1225]
system_reflection_emit_tests!System.Reflection.Emit.Tests.ModuleBuilderDefineEnum.DefineEnum(System.String, System.Reflection.TypeAttributes, System.Type)+0x7a [C:\corefx\src\System.Reflection.Emit\tests\ModuleBuilder\ModuleBuilderDefineEnum.cs @ 37]
```

- Intermittent assert failure at SigPointer::GetTypeVariable (above stacktrace). Let's use this issue to track this intermittent failure.

**Comment by jkotas** [comment]
The intermittent assert failure at `SigPointer::GetTypeVariable` is caused by invalid metadata signature produced by reflection.emit. My guess is that it is likely a silent bad codegen, or GC hole somewhere in the corelib reflection emit code.

**Comment by pgavlin** [comment]
@jkotas: are you looking into the intermittent assert? If not, who should? If you're pretty confident that this is SBCG or a GC hole in the JIT-generated code, I'm happy to take this one.

**Comment by jkotas** [comment]
I am not planning to look into this. I think the first step should be to fix the more deterministic failure with -2146233022 exit code (dotnet/corefx#13238), so that it does not interfere with debugging the less deterministic failure.

**Comment by gkhanna79** [comment]
@karelz who should this be assigned to?

**Comment by karelz** [comment]
Reflection [owners](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#areas) - @AtsushiKan @DnlHarvey, it counts into their area bug count in the reports I'm sending.

**Comment by AtsushiKan** [comment]
>I am not planning to look into this. I think the first step should be to fix the more deterministic failure with -2146233022 exit code (dotnet/corefx#13238, #7669), so that it does not interfere with debugging the less deterministic failure.

That was unblocked by https://github.com/dotnet/coreclr/commit/1de6fe60fd616b4c25f64f8d1c6f17a1e45915d1

Assigning back to @pgavlin since he was offering to look at the intermittent issue.

**Comment by karelz** [comment]
@pgavlin are you actively looking at it now or will be shortly? (just checking)

**Comment by pgavlin** [comment]
@RussKeldorph are we still seeing this?

**Comment by RussKeldorph** [comment]
@pgavlin I think we are not.  We can reopen if we see it again.
